store/rootmulti/store.go

```diff
@@ -63,25 +65,27 @@
 		keysByName:   make(map[string]types.StoreKey),
 		stores:       make(map[types.StoreKey]types.CommitKVStore),
 		listeners:    make(map[types.StoreKey][]types.WriteListener),
+		mtx:          sync.RWMutex{},
 	}
 }
 
 // flush writes all the pending change sets to memiavl tree.
 func (rs *Store) flush() error {
 	var changeSets []*memiavl.NamedChangeSet
-	for key := range rs.stores {
+	rs.mtx.RLock()
```

Comment:
It's easy to deadlock to grab locks in private method, if it's called by another method which already grabbed locks